brainfuck2/bf.c

```diff
@@ -0,0 +1,317 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <libsocket/libinetsocket.h>
+
+#define OP_LIST_DEFAULT_CAPACITY 1 << 5
+#define TAPE_DEFAULT_CAPACITY 1 << 5
+
+#ifdef __clang__
+# define COMPILER "clang"
+#else
+# define COMPILER "GCC"
+#endif
+
+struct op_list;
+
+enum op_type {
+	OP_INC,
+	OP_MOVE,
+	OP_LOOP,
+	OP_PRINT
+};
+
+union op_value {
+	int offset;
+	struct op_list *list;
+};
+
+struct op {
+	enum op_type type;
+	union op_value value;
+};
+
+struct op_list {
+	struct op *ops;
+	int cap, len;
+};
+
+struct op op_new(enum op_type type, union op_value value)
+{
+	struct op op;
+
+	op.type = type;
+	if (type == OP_LOOP)
+		op.value.list = value.list;
+	else
+		op.value.offset = value.offset;
+
+	return op;
+}
+
+void op_list_free(struct op_list *list);
+
+void op_free(struct op op)
+{
+	if (op.type == OP_LOOP) {
+		op_list_free(op.value.list);
+		free(op.value.list);
+	}
+}
+
+void op_list_init(struct op_list *list)
+{
+	list->len = 0;
+	list->cap = OP_LIST_DEFAULT_CAPACITY;
+	list->ops = malloc(sizeof(struct op) * OP_LIST_DEFAULT_CAPACITY);
+}
+
+void op_list_free(struct op_list *list)
+{
+	int i;
+
+	for (i = 0; i < list->len; i += 1)
+		op_free(list->ops[i]);
+
+	free(list->ops);
+}
+
+void op_list_grow(struct op_list *list)
+{
+	struct op *new_list;
+	int i;
+
+	/* double the capacity */
+	list->cap <<= 1;
+	new_list = malloc(sizeof(struct op) * list->cap);
```

Comment:
Please use realloc - it aligns better with the other tests and potentially faster.